### ISSUE TITLE
Evaluate using generics for source/predicate/handler chain

### ISSUE THREAD

**Issue by alvaroaleman** [kubernetes-sigs/controller-runtime#2214]
Predicates currently take a `client.Object` and return a `bool`. With generics, we might be able to have typed generics, which would let ppl avoid a type assertion inside their predicate when they want to access anything other than metadata.

We currently type to `client.Object`, but this is both too much in the case of a `source.Channel` that could take literally everything that might not actually be a `client.Object` and too little in the case of everything else, as we lose the type information.

**Comment by alvaroaleman** [comment]
I did look into this. It works great for just the source/predicate/handler chain, but due to https://github.com/golang/go/issues/49085 it makes the UX of the `builder` terrible, as you can not directly pass a type and handler into `Watches` and friends but instead need to wrap them.

**Comment by nikola-jokic** [comment]
I can give it another try if you'd like :relaxed:

**Comment by alvaroaleman** [comment]
> I can give it another try if you'd like

Its not about giving it a try. I don't think it is possible to do this without that golang issue being solved or accepting that the UX of the builder will become quite a bit worse. Let me try to find the branch I did this in (its been a couple of months) so it can be shown

**Comment by Danil-Grigorev** [comment]
Hi, I added a PR which propagates type information to the handler - https://github.com/kubernetes-sigs/controller-runtime/pull/2698 Also faced with the same issue as @alvaroaleman [mentioned](https://github.com/kubernetes-sigs/controller-runtime/issues/2214#issuecomment-1961380244). One way to tackle this is to pass prepared source.Kind to all builder methods instead of object (`For`, `Owns`, `Watches`), and deprecate, or count non-functioning the `OnlyMetadata` option for the new methods introduced to support source.Kind. Which does not seem to be required to set `PartialObjectMetadata` watches anyway.

Another option is to wrap `client.Object` in a `struct[T]{object T}` (or similar) and pass this around the builder methods instead. Or just make builder methods wrappers and get rid of chaining.

**Comment by alvaroaleman** [comment]
Yeah a wrapper around generic source/predicate/handler is the only way to make this work in the builder.

Just to be clear on this, we will not be merging any code unless we are sure with worsening the builder UX like that.

**Comment by alvaroaleman** [comment]
I guess to extend to the prior message, I would really love to see us using generics there somehow because that would mean that you can build arbitrary controllers that might not even act on kube resources. But not at the price of significantly worsening the experience when building kube controllers, as that will remain the 98% use case of controller-runtime.

**Comment by Danil-Grigorev** [comment]
I looked into this a bit further… The biggest problem with integrating generics is an enormous amount of code duplication, to preserve current API definitions without changes introduced by generics. Essentially everything: predicates, `cache.Informers`, `enqueue mappers`, object handlers, event handlers, `source.Kind` have to have a generic counterpart or just generic version of themselves. Which has a cascade effect on other types across the code.

**TLDR**:  I would personally just prefer having generic predicates and eventHandler, and ensuring the type safety for the objects coming from cache (for starters), and focus on that. WDYT?

The problem I’m facing is again, the builder, or rather the options for each watch input. I’m rotating this pseudocode design in mind at it’s most raw form:

```go
// internal source.Handler:
func NewEventHandler[T any](ctx context.Context, queue workqueue.RateLimitingInterface, 
    handler handler.EventHandler[T], predicates []predicate.Predicate[T]) *EventHandler[T]
```

```go
// handler:
func EnqueueRequestsFromMapFunc[T any](fn MapFunc[T]) EventHandler[T]
```

```go
// Predicate:
func NewPredicateFuncs[T any](filter func(object T) bool) Funcs[T]
```

```go
// Builder:
func (blder *Builder) Add(…) *Builder
func Watch[T any](object T, eventHandler handler.EventHandler[T], opts ...Option[T])
```

```go
// Usage:
predicate := predicate.NewPredicateFuncs(func(pod *corev1.Pod) bool { return true }))
mapHandler := handler.EnqueueRequestsFromMapFunc(func(ctx context.Context, pod *corev1.Pod) []reconcile.Request …
b.Add(Watch(&corev1.Pod{}, mapHandler, predicate))
```
With this approach the first occurrence of `&corev1.Pod` dictates the type for the rest of the declaration, and the predicates or event handlers provided will already expect `*corev1.Pod` as the source object, reducing amount of business logic required to setup a controller, in theory.

But the options do not fit. If I specify type in it, then everything storing the struct to which the option should be applied have to inherit this type, and the `Builder` too, which is incorrect. A controller almost always operates on different objects in tandem.

I think that if the current UX brings enough improvement and it makes sense to add in some (similar or other) for, it would be good to add these changes piece by piece. Refactoring whole source/predicate/handler chain causes changes in the whole project. If we agree on that, we can split it in smaller parts and implement incrementally.

**Comment by alvaroaleman** [comment]
> But the options do not fit. If I specify type in it, then everything storing the struct to which the option should be applied have to inherit this type, and the Builder too, which is incorrect

Exactly, that is the unsolved problem. I do not want to force everyone to use a wrapper before calling `builder.For/Owns/Watches`, as that will lead to an enormous amount of code churn.

**Comment by alvaroaleman** [comment]
Okay so one approach I can think of that would not require changes to exported parts of the builder is to change the `source.Source` interface to not take `Handler` and `Predicates` in `Start`, but get them at creation time. That way, everything generic is inside the Source implementation. This would mean that folks that directly call `controller.Watch` need to change things but not folks that use the builder, which presumably is the majority.

**Comment by Danil-Grigorev** [comment]
I tried that, that works, have a very much WIP implementation for this in this [branch](https://github.com/Danil-Grigorev/controller-runtime/tree/extend-builder). It is simpler than the original form I was trying to pursue, thanks for the pointer.

**Comment by alvaroaleman** [comment]
@Danil-Grigorev would you be up to write a small doc that describes the changes in public interfaces needed for this and the motivation?

**Comment by alvaroaleman** [comment]
I created https://github.com/kubernetes-sigs/controller-runtime/pull/2783 for this which I think is the most compatible way we can do this, let me know what you think.
